bug 701609 - Fix partial install issues.

TimAbraldes opened the pull request:
If the uninstaller is present, rely on it to remove the installation dir. Fixes issue where we inadvertently removed the installation dir out from under the uninstaller, so it failed to run properly. In the callback for icon retrieval, do not proceed if the installation has already aborted, and attemp to remove the installation;r=anant

TimAbraldes committed:
bug 701609 - Fix partial install issues.

If the uninstaller is present, rely on it to remove the installation dir. Fixes issue where we inadvertently removed the installation dir out from under the uninstaller, so it failed to run properly. In the callback for icon retrieval, do not proceed if the installation has already aborted, and attemp to remove the installation;r=anant

mixedpuppy commented on addons/jetpack/lib/nativeshell.js:
```diff
@@ -579,24 +581,22 @@ WinNativeShell.prototype = {
 
-        if (uninstallerFile.exists()) {
+        if (!uninstallerFile.exists()) {
+          this.installDir.remove(true);
```
SCARY!

mixedpuppy replied:
I'll name my next app ../../../

TimAbraldes replied:
this.installDir is built up like so:
      this.installDir = directoryService.get("AppData", Ci.nsIFile);
      this.installDir.append("Mozilla");
      this.installDir.append(this.appName);

According to the docs, we can use nsILocalFile.appendRelativePath which doesn't allow ".." (https://developer.mozilla.org/en/XPCOM_Interface_Reference/nsILocalFile#appendRelativePath%28%29), but that would allow path separators which we also don't want.

It looks like the implementation of nsIFile.append implemented by nsLocalFileWin doesn't allow ".." (https://mxr.mozilla.org/mozilla-central/source/xpcom/io/nsLocalFileWin.cpp?rev=107c941e27dc#1072), and my guess is that nsFile.append is intended to disallow ".."

Maybe we should leave this as is?

mixedpuppy commented on addons/jetpack/lib/nativeshell.js:
```diff
@@ -579,24 +581,22 @@ WinNativeShell.prototype = {
         uninstallerFile.append("uninstall.exe");
 
-        if (uninstallerFile.exists()) {
+        if (!uninstallerFile.exists()) {
```
What if uninstallerFile doesn't exist AND installDir doesn't exist?

TimAbraldes replied:
In that case we can't detect anything for us to remove (I think) so we're done

TimAbraldes committed:
bug 701609 - Avoiding ".." in appName attack.

Using appendRelativePath instead of append.  appendRelativePath does not
allow for ".." to be appended.

TimAbraldes committed:
Revert "bug 701609 - Avoiding ".." in appName attack."

This reverts commit 27640be8297d16dfe77596cfb2679910637d2eba.

It turns out that append doesn't allow ".." and, since it also doesn't allow
path separators, is what we should use.
